server/handlers/project_command_output_handler.go

```diff
@@ -172,40 +184,65 @@ func (p *AsyncProjectCommandOutputHandler) writeLogLine(pull string, line string
 	}
 
 	p.projectOutputBuffersLock.Lock()
-	if p.projectOutputBuffers[pull] == nil {
-		p.projectOutputBuffers[pull] = []string{}
+	if p.projectOutputBuffers[jobID] == nil {
+		p.projectOutputBuffers[jobID] = []string{}
 	}
-	p.projectOutputBuffers[pull] = append(p.projectOutputBuffers[pull], line)
+	p.projectOutputBuffers[jobID] = append(p.projectOutputBuffers[jobID], line)
 	p.projectOutputBuffersLock.Unlock()
 }
 
 //Remove channel, so client no longer receives Terraform output
-func (p *AsyncProjectCommandOutputHandler) Deregister(pull string, ch chan string) {
-	p.logger.Debug("Removing channel for %s", pull)
+func (p *AsyncProjectCommandOutputHandler) Deregister(jobID string, ch chan string) {
+	p.logger.Debug("Removing channel for %s", jobID)
 	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers[pull], ch)
+	delete(p.receiverBuffers[jobID], ch)
 	p.receiverBuffersLock.Unlock()
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(pull string) map[chan string]bool {
-	return p.receiverBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(jobID string) map[chan string]bool {
+	return p.receiverBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(pull string) []string {
-	return p.projectOutputBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(jobID string) []string {
+	return p.projectOutputBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) CleanUp(pull string) {
-	p.projectOutputBuffersLock.Lock()
-	delete(p.projectOutputBuffers, pull)
-	p.projectOutputBuffersLock.Unlock()
+func (p *AsyncProjectCommandOutputHandler) CleanUp(jobContext models.JobContext) {
 
-	// Only delete the pull record from receiver buffers.
-	// WS channel will be closed when the user closes the browser tab
-	// in closeHanlder().
-	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers, pull)
-	p.receiverBuffersLock.Unlock()
+	// Retrieve the jobID from the pullToJobMapping
+	if jobID, ok := p.pullToJobMapping[jobContext]; ok {
```

Comment:
Hmm yeah, this looks promising! I'll give it a shot!

Reply:
So I did some readings about sync Map and it looks like a using a normal map with mutex would be better for our use-case. According to the docs: 

It is optimized for use in concurrent loops with keys that are stable over time, and either few steady-state stores, or stores localized to one goroutine per key.
For use cases that do not share these attributes, it will likely have comparable or worse performance and worse type safety than an ordinary map paired with a read-write mutex.

Since we write to the pullToJobMapping more often and only read from it only once during clean up, I think a normal map would give us better performance overall. Besides, we only read from the pullToJobMapping during clean up after a PR is closed, and I can't think of a scenario where an operation for the PR is still running when the clean up has been initiated, eliminating the need of using mutexes. What do you think?

Reply:
Reference blog: https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c

Reply:
hmm performance is a fair consideration but we don't really care about that as much with atlantis given how long terraform operations can take. I'm moreso trying to optimize for readability here.

Reply:
Yup, I've replaced  it with a sync.Map instead! We could eventually replace all the maps with sync.Map instead if it makes the code more readable!